ucb-rit/coldfront
# Issue 610: Allow superusers to create standalone clusters from UI

**Changes:**
- Adds request wizard for creating standalone clusters. Currently a button to reach this is under the Project page, but this should change.
- Views in `new_project_views/request_views.py`
- Forms in `new_project_forms/request_forms.py`
- Templates under `project/project_request/standalone_cluster` folder
- The wizard asks for the name and description of the cluster -> an existing PI -> if no existing PI provided, a new PI's info -> an existing manager -> if no existing manager provided, a new manager's info -> a review-and-submit page.
- Back-end:
  - Creates a `Resource` object and capitalizes the given name for that object.
  - Creates a `Project` object with the corresponding PI and manager info.

**Testing:**
- Manual testing wizard: 
  - Navigate to `/project/new-standalone-cluster-request/` or click the temporary button "Create new standalone cluster" under the project page.
  - Go through the cluster creation wizard. 
  - To test new PI and/or manager creation, do not select a PI/manager from the dropdowns. The next page will then prompt you to enter user info.
  - After successful submission, go to admin dashboard. Check to see that new Project and Resource objects were created.
    - Check that PI and Manager were created/set correctly under the Project object.
- _TODO Currently doesn't pass_: Testing that a user can join the newly created standalone cluster
  - After creating a standalone cluster via the wizard, a user should be able to request to join that project.
  - Go to `/project/join/` and find the cluster you just made. Clicking the hyperlinked pk currently raises a `NoReverseMatch` error. This is still in progress and I haven't been able to find a fix yet.
    - Supposedly certain backend objects are not being created correctly in the processing part of the form.
    - I have tried adding Allocation, AllocationAttribute, and ProjectTransaction objects to no avail yet.
    - See [docs](https://github.com/ucb-rit/mybrc-mylrc-docs/tree/main/docs/reference/data-modeling), also similar object creation at `coldfront.core.project.management.commands.projects`. (Thanks to Matthew!)

**Other TODO:**
- Should change the name of the Cluster that appears for standalone cluster objects listed in the project join page (`/project/join/`). It currently lists the "Cluster" item as "SAVIO," which is not correct (on BRC build).
